File: hmf/hmf.py
#-*- coding:utf-8 -*-
'''
The MIT License (MIT)

Copyright(C) 2013 <Hooke HU>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
'''
from StringIO import StringIO
from util import *
from tag import *

class Hmf(object):
		""""""
		def __init__(self):
				self._reset()
				
		def _reset(self):
				self.ints = [] #Int pool
				self.strs = [] #str pool
				self.doubles = [] #double pool
				self.bools = [False, True] #bool pool
				self.strem = None
				
		def _set_stream(self, stream):
				self.stream = stream
				
		def write_object(self, obj, stream):
				self._set_stream(stream)
				self._inner_write(obj)
				self._merge_all()
				self._reset()
				
		def _inner_write(self, obj):
				t = type(obj)
				if isinstance(obj, list):
						self._write_array(obj)
				elif isinstance(obj, dict):
						self._write_dict(obj)
				elif isinstance(obj, int):
						self._write_int(obj)
				elif isinstance(obj, float):
						self._write_number(obj)
				elif isinstance(obj, str) or isinstance(obj, unicode):
						self._write_string(obj)
				else:
						print 'error'
						pass
						
		def _write_int(self, v):
				write_varint32(INT_TAG, self.stream)
				if v in self.ints:
						idx = self.ints.index(v)
						write_varint32(idx, self.stream)
						return
				self.ints.append(v)
				idx = len(self.ints) - 1
				write_varint32(idx, self.stream)
				
		def _write_number(self, v):
				write_varint32(DOUBLE_TAG, self.stream)
				if v in self.doubles:
						idx = self.doubles.index(v)
						write_varint32(idx, self.stream)
						return
				self.doubles.append(v)
				idx = len(self.doubles) - 1
				write_varint32(idx, self.stream)
				
		def _write_string(self, v):
				write_varint32(STRING_TAG, self.stream)
				if v in self.strs:
						idx = self.strs.index(v)
						write_varint32(idx, self.stream)
						return
				
				self.strs.append(v)
				idx = len(self.strs) - 1
				write_varint32(idx, self.stream)
				
		def _write_array(self, v):
				#write array tag to stream
				write_varint32(ARRAY_TAG, self.stream)
				l = len(v)
				write_varint32(l, self.stream)
				for i in v:
						self._inner_write(i)
				
		def _write_dict(self, v):
				#write object tag to stream
				write_varint32(OBJECT_TAG, self.stream)
				l = len(v)
				write_varint32(l, self.stream)
				for k1, v1 in v.items():
						self._inner_write(k1)
						self._inner_write(v1)
						
		def _merge_all(self):
				rst = StringIO()
				write_varint32(len(self.ints), rst)
				for i in self.ints:
						write_varint32(i, rst)
				write_varint32(len(self.doubles), rst)
				for i in self.doubles:
						write_double(i, rst)
				write_varint32(len(self.strs), rst)
				for i in self.strs:
						write_varint32(len(i), rst)
						write_str(i, rst)
				self.stream.seek(0)
				s = self.stream.read()
				self.stream.seek(0)
				rst.seek(0)
				self.stream.write(rst.read())
				self.stream.write(s)
				
		def read_object(self, stream):
				self._set_stream(stream)
				self._init_pool()
				tag = read_varint32(self.stream)
				rst = None
				if tag == ARRAY_TAG:
						rst = self._read_array()
						self._reset()
						return rst
				elif tag == OBJECT_TAG:
						rst = self._read_dict()
						self._reset()
						return rst
				else:
						self._reset()
						return None
						
		def _read_dict(self):
				d = {}
				l = read_varint32(self.stream)
				for i in range(l):
						tag = read_varint32(self.stream)
						if tag == ARRAY_TAG:
								k = self._read_array()
						elif tag == OBJECT_TAG:
								k = self._read_dict()
						elif tag == INT_TAG:
								k = self.ints[read_varint32(self.stream)]
						elif tag == STRING_TAG:
								k = self.strs[read_varint32(self.stream)]
						elif tag == DOUBLE_TAG:
								k = self.doubles[read_varint32(self.stream)]
								
						tag = read_varint32(self.stream)
						if tag == ARRAY_TAG:
								v = self._read_array()
						elif tag == OBJECT_TAG:
								v = self._read_dict()
						elif tag == INT_TAG:
								v = self.ints[read_varint32(self.stream)]
						elif tag == STRING_TAG:
								v = self.strs[read_varint32(self.stream)]
						elif tag == DOUBLE_TAG:
								v = self.doubles[read_varint32(self.stream)]
						print k
						d[k] = v
				return d
				
		def _read_array(self):
				rst = []
				l = read_varint32(self.stream)
				for i in range(l):
						tag = read_varint32(self.stream)
						if tag == ARRAY_TAG:
								rst.append(self.read_array())
						elif tag == OBJECT_TAG:
								rst.append(self.read_dict())
						elif tag == INT_TAG:
								rst.append(self.ints[read_varint32(self.stream)])
						elif tag == STRING_TAG:
								rst.append(self.strs[read_varint32(self.stream)])
						elif tag == DOUBLE_TAG:
								rst.append(self.strs[read_varint32(self.stream)])
				return rst
				
		def _init_pool(self):
				len_ints = read_varint32(self.stream)
				for i in range(len_ints):
						self.ints.append(read_varint32(self.stream))
				len_doubles = read_varint32(self.stream)
				for i in range(len_doubles):
						self.doubles.append(read_double(self.stream))
				len_strs = read_varint32(self.stream)
				for i in range(len_strs):
						l = read_varint32(self.stream)
						self.strs.append(read_str(self.stream, l))
				
if __name__ == "__main__":
		import time, pyamf
		h = Hmf()
		o = {'a':[1,2,3], 'b':{'c':1}, 1:2, 3:0.3312344}
		print o
		stream = StringIO()
		h.write_object(o, stream)
		stream.seek(0)
		e = h.read_object(stream)
		print e
		print "success"
		
		'''
		fd = open("E://out.hmf", "rb")
		st = StringIO()
		st.write(fd.read())
		fd.close()
		st.seek(0)
		print time.time()
		e = h.read_object(st)
		print len(e)
		fd.close()'''
		

File: MogoSolution/LoaderLib/Libs/HMF/Hmf.cs
/**
The MIT License (MIT)

Copyright(C) 2013 <Hooke HU>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
**/
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace HMF
{
    public class Hmf
    {
        private List<int> ints = new List<int>();
        private List<double> doubles = new List<double>();
        private List<bool> bools = new List<bool>();
        private List<string> strs = new List<string>();

        private Stream stream = null;

        public Hmf()
        {
        }

        private void Reset()
        {
            ints = new List<int>();
            doubles = new List<double>();
            strs = new List<string>();
            bools = new List<bool>();
            bools.Add(false);
            bools.Add(true);
            stream = null;
        }

        private void SetStream(Stream stream)
        {
            this.stream = stream;
        }

        public void WriteObject(object obj, Stream stream)
        {
            SetStream(stream);
            RealWrite(obj);
            MergeAll();
        }

        private void RealWrite(object obj)
        {
            if(obj.GetType() == typeof(int))
            {
                WriteInt((int)obj);
            }
            else if(obj.GetType() == typeof(List<object>))
            {
                WriteArray((List<object>)obj);
            }
            else if(obj.GetType() == typeof(Dictionary<object, object>))
            {
                WriteDict((Dictionary<object, object>)obj);
            }
            else if(obj.GetType() == typeof(double))
            {
                WriteDouble((double)obj);
            }
            else if(obj.GetType() == typeof(string))
            {
                WriteString((string)obj);
            }
        }

        private void WriteInt(int v)
        {
            Util.WriteVarint((int)Tag.INT_TAG, stream);
            int idx = ints.IndexOf(v);
            if(idx != -1)
            {
                Util.WriteVarint(idx, stream);
                return;
            }
            ints.Add(v);
            idx = ints.Count - 1;
            Util.WriteVarint(idx, stream);
        }

        private void WriteDouble(double v)
        {
            Util.WriteVarint((int)Tag.DOUBLE_TAG, stream);
            int idx = doubles.IndexOf(v);
            if (idx != -1)
            {
                Util.WriteVarint(idx, stream);
                return;
            }
            doubles.Add(v);
            idx = doubles.Count - 1;
            Util.WriteVarint(idx, stream);
        }

        private void WriteString(string v)
        {
            Util.WriteVarint((int)Tag.STRING_TAG, stream);
            int idx = strs.IndexOf(v);
            if (idx != -1)
            {
                Util.WriteVarint(idx, stream);
                return;
            }
            strs.Add(v);
            idx = strs.Count - 1;
            Util.WriteVarint(idx, stream);
        }

        private void WriteArray(List<object> v)
        {
            Util.WriteVarint((int)Tag.ARRAY_TAG, stream);
            int len = v.Count;
            Util.WriteVarint(len, stream);
            for (int i = 0; i < len; i++)
            {
                RealWrite(v[i]);
            }
        }

        private void WriteDict(Dictionary<object, object> v)
        {
            Util.WriteVarint((int)Tag.OBJECT_TAG, stream);
            int len = v.Count;
            Util.WriteVarint(len, stream);
            foreach (var i in v)
            {
                RealWrite(i.Key);
                RealWrite(i.Value);
            }
        }

        private void MergeAll()
        {
            MemoryStream st = new MemoryStream();
            Util.WriteVarint(ints.Count, st);
            for(int i = 0; i < ints.Count; i++)
            {
                Util.WriteVarint(ints[i], st);
            }
            Util.WriteVarint(doubles.Count, st);
            for(int i = 0; i < doubles.Count; i++)
            {
                Util.WriteDobule(doubles[i], st);
            }
            Util.WriteVarint(strs.Count, st);
            for(int i = 0; i < strs.Count; i++)
            {
                Util.WriteVarint(strs[i].Length, st);
                Util.WriteStr(strs[i], st);
            }
            stream.Position = 0;
            st.Position = 0;
            int stLen = (int)st.Length;
            int streamLen = (int)st.Length;
            byte[] stBytes = new byte[stLen];
            byte[] streamBytes = new byte[streamLen];
            stream.Read(streamBytes, 0, streamLen);
            st.Read(stBytes, 0, stLen);
            stream.Position = 0;
            stream.Write(stBytes, 0, stLen);
            stream.Write(streamBytes, 0, streamLen);
        }

        public object ReadObject(Stream stream)
        {
                SetStream(stream);
                InitPool();
#if UNITY_IPHONE
#else
                System.Console.WriteLine("p " + stream.Position);
#endif
                byte tag = (byte)Util.ReadVarint(stream);
                object rst = null;
                if(tag == Tag.ARRAY_TAG)
                {
                    rst = ReadArray();
                    Reset();
                    return rst;
                }
                else if(tag == Tag.OBJECT_TAG)
                {
                    rst = ReadDict();
                    Reset();
                    return rst;
                }
#if UNITY_IPHONE
#else
                System.Console.WriteLine("fuck " + tag);
#endif
                Reset();
                return null;
        }

        private void InitPool()
        {
            int len_ints = Util.ReadVarint(stream);
            for(int i = 0; i < len_ints; i++)
            {
                this.ints.Add(Util.ReadVarint(stream));
            }
            int len_doubles = Util.ReadVarint(stream);
            for (int i = 0; i < len_doubles; i++)
            {
                this.doubles.Add(Util.ReadDouble(stream));
            }
            int len_strs = Util.ReadVarint(stream);
            for(int i = 0; i < len_strs; i++)
            {
                int l = Util.ReadVarint(stream);
                this.strs.Add(Util.ReadStr(l, stream));
            }
        }

        private List<object> ReadArray()
        {
            List<object> rst = new List<object>();
            int len = Util.ReadVarint(stream);
            for(int i = 0; i < len; i++)
            {
                byte tag = (byte)Util.ReadVarint(stream);
                if(tag == Tag.ARRAY_TAG)
                {
                    rst.Add(ReadArray());
                }
                else if(tag == Tag.OBJECT_TAG)
                {
                    rst.Add(ReadDict());
                }
                else if(tag == Tag.INT_TAG)
                {
                    rst.Add(ints[Util.ReadVarint(stream)]);
                }
                else if(tag == Tag.STRING_TAG)
                {
                    rst.Add(strs[Util.ReadVarint(stream)]);
                }
                else if(tag == Tag.DOUBLE_TAG)
                {
                    rst.Add(strs[Util.ReadVarint(stream)]);
                }

            }
            return rst;
        }

        private Dictionary<object, object> ReadDict()
        {
            Dictionary<object, object> rst = new Dictionary<object, object>();
            int len = Util.ReadVarint(stream);
            for (int i = 0; i < len; i++)
            {
                byte tag = (byte)Util.ReadVarint(stream);
                object k = null;
                if(tag == Tag.ARRAY_TAG)
                {
                    k = ReadArray();
                }
                else if(tag == Tag.OBJECT_TAG)
                {
                    k = ReadDict();
                }
                else if(tag == Tag.INT_TAG)
                {
                    k = ints[Util.ReadVarint(stream)];
                }
                else if(tag == Tag.STRING_TAG)
                {
                    k = strs[Util.ReadVarint(stream)];
                }
                else if(tag == Tag.DOUBLE_TAG)
                {
                    k = doubles[Util.ReadVarint(stream)];
                }

                tag = (byte)Util.ReadVarint(stream);
                object v = null;
                if (tag == Tag.ARRAY_TAG)
                {
                    v = ReadArray();
                }
                else if (tag == Tag.OBJECT_TAG)
                {
                    v = ReadDict();
                }
                else if (tag == Tag.INT_TAG)
                {
                    v = ints[Util.ReadVarint(stream)];
                }
                else if (tag == Tag.STRING_TAG)
                {
                    v = strs[Util.ReadVarint(stream)];
                }
                else if (tag == Tag.DOUBLE_TAG)
                {
                    v = doubles[Util.ReadVarint(stream)];
                }
                rst.Add(k, v);

            }
            return rst;
        }
    }
}
